### PR TITLE
ENH: Scale-dependent slope and curvature

### PR DESCRIPTION
**Pull request by pastewka**
This PR implements pipeline functions to compute scale-dependent slope and curvature from the ACF. Other fixes:
- Windowing is now a pipeline filter function
- BUG: RMS slope/curvature calculation with a cutoff requires the prior application of a window

### CODE REVIEW

File: SurfaceTopography/Generic/Curvature.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    B = 8 * A[1:n] - 2 * A[2::2]
+    nz = np.nonzero(B < 0)[0]
+    if len(nz) > 0:
+        n = nz[0]
+    # Important: The following expression relies on the fact that r is equally spaced!
+    return r[:n], np.sqrt(B[:n]) / r[:n] ** 2
+
+
+def scale_dependent_curvature_2D(topography, nbins=None):
+    r"""
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/Generic/Curvature.py:76]: same here.

File: SurfaceTopography/Uniform/Autocorrelation.py [hunk 2]
```diff
@@ -79,8 +81,20 @@ def autocorrelation_1D(topography, direction=0):
...
+    if subinterval is not None:
+        # If a subinterval is selected, the topography is automatically
+        # non-periodic
+        is_periodic = False
+        l, r = subinterval
+        p = p[l:r]
+        # Compute new sizes
+        dx = sx/nx
+        nx = len(p)
+        print(nx, l-r)
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/Uniform/Autocorrelation.py:94]: debug bprint

File: SurfaceTopography/Uniform/Autocorrelation.py [hunk 3]
```diff
@@ -128,17 +142,26 @@ def autocorrelation_1D(topography, direction=0):
...
+        Number of bins for radial average. Bins are automatically determined
+        if set to None. (Default: None) Note: Returned array can be smaller
+        than this because bins without data points are discarded.
+        (Default: None)
+    bin_edges : {'log', 'quadratic', 'linear', array_like}, optional
+        Edges used for binning the average. Specifying 'log' yields bins
+        equally spaced on a log scale, 'quadratic' yields bins with
+        similar number of data points and 'linear' yields linear bins.
+        Alternatively, it is possible to explicitly specify the bin edges.
+        If `bin_edges` are explicitly specified, then `rmax` and `nbins` is
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/Uniform/Autocorrelation.py:163]: you are mentioning arguments that are not in the signature

Comment (AntoineSIMTEK) [reply SurfaceTopography/Uniform/Autocorrelation.py:163]: I guess this is a copy paste error from the "radial_average" function

File: SurfaceTopography/Uniform/Filtering.py [hunk 4]
```diff
@@ -23,10 +23,122 @@
...
+    @property
+    def window_data(self):
+        if self._window_data is None:
+            self._make_window()
+        return self._window_data
+
+    def heights(self):
+        """ Computes the windowed topography.
+        """
+        if self._window_data is None:
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/Uniform/Filtering.py:128]: I think you can replace this statement with one: if self.window_data() is None

File: SurfaceTopography/Uniform/PowerSpectrum.py [hunk 5]
```diff
@@ -114,9 +100,18 @@ def power_spectrum_2D(topography, nbins=100,  # pylint: disable=invalid-name
...
+    nbins : int, optional
+        Number of bins for radial average. Bins are automatically determined
+        if set to None. (Default: None) Note: Returned array can be smaller
+        than this because bins without data points are discarded.
+        (Default: None)
+    bin_edges : {'log', 'quadratic', 'linear', array_like}, optional
+        Edges used for binning the average. Specifying 'log' yields bins
+        equally spaced on a log scale, 'quadratic' yields bins with
+        similar number of data points and 'linear' yields linear bins.
+        Alternatively, it is possible to explicitly specify the bin edges.
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/Uniform/PowerSpectrum.py:112]: Here again rmax and nbins should not be mentioned here

File: SurfaceTopography/common.py [hunk 6]
```diff
@@ -30,21 +30,35 @@
...
+
     Parameters
     ----------
     C_xy : array_like
         2D-array of values to be averaged.
-    rmax : float
-        Maximum radius.
-    nbins : int
-        Number of bins for averaging.
+    rmax : float, optional
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/common.py:44]: I guess it is automatically determined if not provided.

**Review by AntoineSIMTEK — CHANGES_REQUESTED** [review]
We have a function rms_slope that can also compute a kind of scale dependent rms_slope by providing a filter. 

And we have a function called `scale_dependent_rms_slope` that computes it using the autocorrelation function.

But I think I have no better idea for the denomination

**Comment by pastewka** [comment]
> We have a function rms_slope that can also compute a kind of scale dependent rms_slope by providing a filter.
> 
> And we have a function called `scale_dependent_rms_slope` that computes it using the autocorrelation function.
> 
> But I think I have no better idea for the denomination

I agree. We need to cleanup the whole API at some point. Another problem is: `rms_height` has an argument that allows us to tell it whether to do the 1D or 2D calculations, `rms_slope` and `rms_curvature` do not. For `autocorrelation` and `power_spectrum` there are explicit 1D and 2D versions. I'll create an issue for this.

**Comment by pastewka** [comment]
Thanks for reviewing this @AntoineSIMTEK - I should have addressed all issues...

File: SurfaceTopography/Generic/Curvature.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    Compute the two-dimensional, radially averaged scale-dependent curvature.
+
+    The scale-dependent curvature is given by
+
+       .. math::
+         :nowrap:
+
+         \begin{equation}
+         h_\text{rms}^\prime(\lambda) = \left[8A(\lambda) - 2A(2\lambda)\right]^{1/2}/\lambda
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/Generic/Curvature.py:85]: I think you are still missing one prime and a square on the lambda

**Review by AntoineSIMTEK — CHANGES_REQUESTED** [review]

File: SurfaceTopography/Generic/Curvature.py (hunk 7 above)
Comment (pastewka) [reply SurfaceTopography/Generic/Curvature.py]: Updated this!

**Review by AntoineSIMTEK — APPROVED** [review]

**Pull request merged**